### ISSUE TITLE
efi-stub: Load extensions and credentials from more locations

### ISSUE THREAD

**Issue by AdrianVovk** [systemd/systemd#21289]
**TL;DR:** systemd-stub should load initramfs extensions from a kernel-agnostic location like `\systemd-creds\*.cred`. Also, instead of throwing out the unsafe cmdline provided by the bootloader, it should package it up and provide it at `/.extra/unsafe-cmdline` in the initramfs.

**Is your feature request related to a problem? Please describe.**
Given [@poettering's recent blog post](https://0pointer.net/blog/authenticated-boot-and-disk-encryption-on-linux.html), I've started [some initial planning work](https://gitlab.com/carbonOS/build-meta/-/issues/44) on getting this implemented for my ostree-based distro. Here's two big use cases I've come across, both involving getting data into the initramfs:

1) OSTree uses type 1 BLS config. It's not going to use type 2 without massive rewrites, and not only to the bootloader-specific code. OSTree relies on the BLS fragments to enumerate deployments, for instance. Anyway, that's fine: from my understanding, I can still package up the kernel + initramfs + cmdline with the efi-stub and sign it, and then just have a type 1 BLS config point to it. However, here's the issue: ostree needs to pass the hash of the deployment into the initramfs. It currently does this via a `ostree=HASH` kernel command line argument. Each BLS config entry has a different hash. The initramfs then uses the hash to get the right tree from the OSTree repository and boot into it. When using the stub, the kernel arguments passed in from systemd-boot will get ignored (GOOD!), but this also includes the required `ostree=` argument (BAD!). Kernel images can be—and often are—shared across deployments, so dropping files next to the kernel won't help. Thus, there needs to be some way to pass data from the type 1 BLS fragment into the initramfs
2) Some of the credentials that the boot chain will depend on don't really depend on kernel or initramfs, and are really tied to the specific machine. Thus, it doesn't make sense to place them next to the kernel. For example, the credential (decryptable by TPM + recovery key) that contains the key used by dm-integrity. That way if the TPM fails/whatever the recovery key could be used to unlock the dm-integrity for the rootfs. Hopefully I'm not woefully misunderstanding how dm-integrity works here; documentation is pretty sparse

**Describe the solution you'd like**
1) I think the simplest solution would look something like this. When secure-boot is on, instead of throwing out the bootloader-provided command line, package it up into a CPIO archive and ship it like you do with everything else in `foo.efi.extra.d`. Then my initramfs could read `/proc/cmdline`, see that `ostree=` is missing, and fall back to looking in `/.extra/unsafe-cmdline` or so. When secure-boot is off, the unsafe cmdline would replace the image's baked-in cmdline, so my initramfs would just read `/proc/cmdline`, find `ostree=`, and boot like normal. A more involved solution could be to add a new key to BLS type 1 fragments (i.e. `credential foo=BASE64`) that gets passed to the stub somehow and inserted into the initramfs.
2) The simplest solution here is probably just adding a drop-in dir that's unrelated to the image name that the stub loads. I propose (where `\` = the root of the ESP or whatever storage medium the stub is being loaded from) `\systemd-cred\*.cred`, or `\loader\creds\*.cred`, or something similar. Then these folders would get populated at install time with the necessary credentials

**Describe alternatives you've considered**
For use case 1, I guess OSTree could be heavily restructured to use type 2 BLS config. Then ostree can drop a kernel  per deployment into the right dir, name it after the deployment hash, and then drop necessary files into `HASH.efi.extra.d`. That way it's still atomic if it creates files in the right order. However, I honestly don't see ostree ever doing this, as it would require massive rewrites. Also, since vfat doesn't support links and deployments will often share kernels, duplicate kernels will be floating around the ESP.
For use case 2, I don't really see a reasonable alternative solution. Copying the credentials on each kernel update? That seems wasteful.

**The systemd version you checked that didn't have the feature you are asking for**
v249, git (unless I somehow missed a PR!)

**Comment by poettering** [comment]
Not sure i follow. If you care about security why would it be safe to use an untrusted kernel cmdline? If it's not verified it cannot be trusted...

If ostree wants to use the same kernel with multiple rootfs subdirs, why not encode the rootfs subdir to use in some credential? credentials can be TPM encrypted/authenticated, so that way you could trust it?

I think this is something to discuss with the ostree people.

**Comment by AdrianVovk** [comment]
> If you care about security why would it be safe to use an untrusted kernel cmdline?

Well, you can't really modify the `ostree=` arg to do anything malicious. But I'd be wrapping it in a credential anyway just to be extra safe. The rest of the unsafe kernel cmdline won't get used at all, which is the point of putting it in `/.extra` and parsing it manually in the initramfs. 

> why not encode the rootfs subdir to use in some credential? credentials can be TPM encrypted/authenticated, so that way you could trust it?

I would be doing it that way! I just don't know how to pass the encoded credential's data from the BLS fragment to the initramfs. If the unsafe cmdline stuff were actually implemented, I'd pass the credential through there, then extract & decrypt & parse it from userspace in the initramfs. If you'd prefer, I guess systemd-stub itself can parse the unsafe cmdline, pull out all credentials, and drop them into `/.extra`. Then the initramfs can use them like any other credential. I proposed it the other way because it would require less code to be added to the stub

I don't see a better mechanism to pass data from the BLS fragment to the initramfs. Maybe you have a better idea?

> I think this is something to discuss with the ostree people.

Of course! This will need work on the ostree side too. However, systemd is missing features that would make this impossible to implement except for a massive rewrite of ostree's deployment code. So I opened the issue here first to figure out what's best for systemd.

P.S. what are your thoughts on defining a generic directory to put credentials in that will get loaded by any kernel? You didn't comment on it

**Comment by poettering** [comment]
> > If you care about security why would it be safe to use an untrusted kernel cmdline?
> 
> Well, you can't really modify the `ostree=` arg to do anything malicious. 

You can't? I though that's where you specify the top-level hash, no? i.e. it's versioning, not? If so it's usable for downgrade attacks, no? What am I missing?



> I would be doing it that way! I just don't know how to pass the encoded credential's data from the BLS fragment to the initramfs

Why do you want to put it in BLS? Just use sd-stub and put it next to the kernel image?

**Comment by AdrianVovk** [comment]
> If so it's usable for downgrade attacks, no?

Basically the only possible valid input for `ostree=` will be the options you have available at the bootloader. So the attacker could just as easily press and hold the spacebar at boot and boot the older version from the menu. No other hashes will successfully boot the system. That's just a side effect of the way OSTree does its bookkeeping

I'm still working on figuring out how to prevent downgrade attacks while preserving downgrades. Perhaps I measure the deployment hash into the TPM so that booting a different hash throws out the key to /var. And I make both hashes valid during boot counting, so failed upgrades and automatic downgrades are fully seamless. Then once boot counting clears the deployment as good, I throw out the old hash and thus a downgrade will require the recovery password.

But again, this is work for my distro and not for systemd. I just need the data in the initramfs and I can make it secure from there

> Why do you want to put it in BLS? Just use sd-stub and put it next to the kernel image?

In OSTree, each deployment is, by definition, one BLS fragment. So each BLS fragment *will* have a different `ostree=` value. They could and often do share a kernel, however. If I put the `ostree=` arg next to the kernel, then two deployments (aka two BLS fragments), will have the same `ostree=` arg and OSTree breaks.

**Comment by poettering** [comment]
> > If so it's usable for downgrade attacks, no?
> 
> Basically the only possible valid input for `ostree=` will be the options you have available at the bootloader. So the attacker could just as easily press and hold the spacebar at boot and boot the older version from the menu. No other hashes will successfully boot the system. That's just a side effect of the way OSTree does its bookkeeping.

Security through interactivity is an acceptable concept in many cases. i.e. if the user has to explicitly do something interactively on the keyboard things are fine that aren't fine if you can trigger them from code that doesn't require interactivity.

note that ostree isn#t really designed with security in mind, I doubt it makes sense to shoehorn security into it really. The fact that it doesn't validate on access and trusts the fs is problematic, as there is no offline security.

> I'm still working on figuring out how to prevent downgrade attacks while preserving downgrades. Perhaps I measure the deployment hash into the TPM so that booting a different hash throws out the key to /var. And I make both hashes valid during boot counting, so failed upgrades and automatic downgrades are fully seamless. Then once boot counting clears the deployment as good, I throw out the old hash and thus a downgrade will require the recovery password.

I think downgrades through user interaction are fine. Automated downgrades are not.

> > Why do you want to put it in BLS? Just use sd-stub and put it next to the kernel image?
> 
> In OSTree, each deployment is, by definition, one BLS fragment. So each BLS fragment _will_ have a different `ostree=` value. They could and often do share a kernel, however. If I put the `ostree=` arg next to the kernel, then two deployments (aka two BLS fragments), will have
the same `ostree=` arg and OSTree breaks.

If you use the boot loader spec type #1 entries you can't sign the combination you are booting. Use boot loader spec type #2 entries (i.e. unified kernels) so that you can sign the whole shebang in one. 

The best you can do with type #1 entries is tpm measure stuff, but that's a much weaker model.

hwo do you sign your initrd anyway if you don't use unified kernels?

**Comment by poettering** [comment]
ah i get it now, you want to sign kernel+initrd, but leave the ostree= out of it and accept it entirely untrusted?

that sounds like a weird model to me... the hash is the version of the OS you want to boot into, and that's kinda the most interesting thing to protect here: giving attackers the ability to combine arbitrary kernels with arbitrary userspaces is problematic given you don#t even have any verification that a combination is ok, not even interactive...

that said, i think it would be ok to read creds from a fixed dir in the ESP in addition to the per-kernel one. send a patch.

**Comment by AdrianVovk** [comment]
> The fact that it doesn't validate on access and trusts the fs is problematic, as there is no offline security.

I'll be using dm-integrity to verify that the rootfs hasn't been tampered with offline before mounting. 

> I think downgrades through user interaction are fine. Automated downgrades are not.

Automated downgrades are a must; What if the user upgrades to a new version and their system doesn't boot. I 100% expect the OS to seamlessly downgrade itself in this case. This is how Chrome OS works also. It's a good UX. Again, outside of bootcounting, there shouldn't be a need for automated downgrades. But I feel this is a bit off topic

> ah i get it now, you want to sign kernel+initrd, but leave the ostree= out of it and accept it entirely untrusted?

As I've said earlier, I could encrypt the ostree= arg via TPM to verify it. However, yes that's the general idea. Sign kernel+initrd+**default cmdline**, and add ostree= on top of it

> giving attackers the ability to combine arbitrary kernels with arbitrary userspaces is problematic given you don#t even have any verification that a combination is ok, not even interactive

Good point. I'll have to add that to my list of things to protect against. A way around this issue that I quickly thought up is to burn a UUID into the kernel image @ build and sign time, and when I'm binding together a kernel + userspace at update time I drop the same UUID into userspace. Then the initramfs will refuse to boot a userspace without the UUID baked into it. dm-integrity will prevent an attacker from adding in the UUID to the userspace offline.

> that said, i think it would be ok to read creds from a fixed dir in the ESP in addition to the per-kernel one. send a patch.

:+1:

**Comment by AdrianVovk** [comment]
Usecase 2 was merged. I think I'll close this for now, and reopen/make a new issue if I end up needing usecase 1. I might have a way to work around it

**Comment by medhefgo** [comment]
> Usecase 2 was merged. I think I'll close this for now, and reopen/make a new issue if I end up needing usecase 1. I might have a way to work around it

I'm just gonna assume you hit the wrong button. Feel free to reopen.
